experimental/lib/Support/BooleanLogic/Parser.cpp

```diff
@@ -0,0 +1,215 @@
+      Compare::greaterThan},
+     {Compare::lessThan, Compare::lessThan, Compare::lessThan,
+      Compare::lessThan, Compare::equal, Compare::lessThan, Compare::error},
+     {Compare::greaterThan, Compare::greaterThan, Compare::greaterThan,
+      Compare::error, Compare::greaterThan, Compare::error,
+      Compare::greaterThan},
+     {Compare::greaterThan, Compare::greaterThan, Compare::greaterThan,
+      Compare::error, Compare::greaterThan, Compare::error,
+      Compare::greaterThan},
+     {Compare::lessThan, Compare::lessThan, Compare::lessThan,
+      Compare::lessThan, Compare::error, Compare::lessThan, Compare::accept}}};
+
+// returns a dynamically-allocated variable
+BoolExpression *dynamatic::experimental::boolean::constructNodeOperator(
+    StackNode *operate, StackNode *s1, StackNode *s2) {
+  assert(operate != nullptr && s1 != nullptr && s2 != nullptr);
+  Token t = operate->term;
+  ExpressionType oo;
+  if (t.tokenType == TokenType::AndToken)
+    oo = ExpressionType::And;
+  else
+    oo = ExpressionType::Or;
+  BoolExpression *e1 = s1->expr;
+  BoolExpression *e2 = s2->expr;
+  return new Operator(oo, e1, e2);
+}
+
+// returns a dynamically-allocated variable
+BoolExpression *
+dynamatic::experimental::boolean::constructNodeNegator(StackNode *s1) {
+  assert(s1 != nullptr);
+  return new Operator(ExpressionType::Not, nullptr, s1->expr);
+}
+
+// returns a dynamically-allocated variable
+StackNode *dynamatic::experimental::boolean::termToExpr(StackNode *s) {
+  assert(s != nullptr);
+  ExpressionType t = ExpressionType::Variable;
+  if (s->term.lexeme == "0")
+    t = ExpressionType::Zero;
+  if (s->term.lexeme == "1")
+    t = ExpressionType::One;
+  return new StackNode(StackNodeType::Expr, new SingleCond(t, s->term.lexeme));
+}
+
+// returns a dynamically-allocated variable
+StackNode *dynamatic::experimental::boolean::constructTermStackNode(Token t) {
+  return new StackNode(StackNodeType::Term, std::move(t));
+}
+
+// returns a dynamically-allocated variable
+StackNode *dynamatic::experimental::boolean::constructOperatorStackNode(
+    StackNode *operate, StackNode *s1, StackNode *s2) {
+  return new StackNode(StackNodeType::Expr,
+                       constructNodeOperator(operate, s1, s2));
+}
+
+// returns a dynamically-allocated variable
+StackNode *
+dynamatic::experimental::boolean::constructNegatorStackNode(StackNode *s1) {
+  return new StackNode(StackNodeType::Expr, constructNodeNegator(s1));
+}
+
+StackNode *
+dynamatic::experimental::boolean::reduce(std::stack<StackNode *> stack) {
+  // Case 1: Handling parentheses: expr --> ( expr )
+  if (stack.size() == 3 &&
+      (stack.top()->type == StackNodeType::Term &&
+       stack.top()->term.tokenType == TokenType::LparenToken)) {
+    stack.pop();
+    StackNode *ex = stack.top();
+    stack.pop();
+    if (stack.top()->type == StackNodeType::Term &&
+        stack.top()->term.tokenType == TokenType::RparenToken)
+      return ex;
+    syntaxError();

```

Comment:
It would be much better to print a different message than the generic one for syntax errors in this case. This would help users find the issue in their string. This comment applies to other places where you use `syntaxError()`.